src/frontend/src/flows/authenticate/validateDerivationOrigin.ts

```diff
@@ -0,0 +1,76 @@
+import { Principal } from "@dfinity/principal";
+
+export type ValidationResult =
+  | { result: "valid" }
+  | { result: "invalid"; message: string };
+
+export const validateDerivationOrigin = async (
+  authRequestOrigin: string,
+  derivationOrigin?: string
+): Promise<ValidationResult> => {
+  if (
+    derivationOrigin === undefined ||
+    derivationOrigin === authRequestOrigin
+  ) {
+    // this is the default behaviour -> no further validation necessary
+    return { result: "valid" };
+  }
+
+  // check format of derivationOrigin
+  const matches = /^https:\/\/([\w-]*)(\.raw)?\.ic0\.app$/.exec(
+    derivationOrigin
+  );
+  if (matches === null) {
+    return {
+      result: "invalid",
+      message:
+        'derivationOrigin does not match regex "^https:\\/\\/([\\w-]*)(\\.raw)?\\.ic0\\.app$"',
+    };
+  }
+
+  try {
+    const canisterId = Principal.fromText(matches[1]); // verifies that a valid canister id was matched
```

Comment:
does this mean that all valid principals are valid canister IDs?

Reply:
No, but only valid principals could potentially be canister ids. If the matched value is not even a principal it is certainly not a valid canister id. This is here to give developers early (and more specific) feedback if they have a malformed `derivationOrigin` (e.g. due to a typo).

I agree, the comment is misleading. I will update it when I revisit this code.